### PR TITLE
Update requests to fix CVEs (security)

### PR DESCRIPTION
**Pull request by jk464**
Fixes up a host of CVEs in the `st2` package:

Note: The `XRAY` references are vulnerabilities listed by [JFrog Xray](https://jfrog.com/help/r/get-started-with-the-jfrog-platform/jfrog-xray), that don't seem to have a corresponding CVE. JFrog doesn't seem to publish these references publicly - but I've linked to the issue disclosing the vulnerability thats referenced by the `XRAY` entry.

# Bump cryptography to 41.0.4, pyopenssl to 23.2.0

Fixes:
* [CVE-2023-4807](https://nvd.nist.gov/vuln/detail/CVE-2023-4807)
* [CVE-2023-2650](https://nvd.nist.gov/vuln/detail/CVE-2023-2650)
* [CVE-2023-3446](https://nvd.nist.gov/vuln/detail/CVE-2023-3446)

pyopenssl 23.2.0 required for cryptography to 41.0.x support

# Bump virtualenv to 20.16.7

Fixes:
* [CVE-2019-20916](https://nvd.nist.gov/vuln/detail/CVE-2019-20916)

# Bump importlib-metadata to 4.10.1

Fixes:
* [XRAY-195083](https://github.com/python/importlib_metadata/issues/361)

# Bump requests to 2.31.0

Fixes:
* [CVE-2023-32681](https://nvd.nist.gov/vuln/detail/CVE-2023-32681)

# Bump gitpython to 3.1.37

Fixes:
* [CVE-2023-40267](https://nvd.nist.gov/vuln/detail/CVE-2023-40267)
* [CVE-2023-41040](https://nvd.nist.gov/vuln/detail/CVE-2023-41040)
* [CVE-2023-40590](https://nvd.nist.gov/vuln/detail/CVE-2023-40590)
* [CVE-2022-24439](https://nvd.nist.gov/vuln/detail/CVE-2022-24439)
* [XRAY-198950](https://huntr.com/bounties/8549d81f-dc45-4af7-9f2a-2d70752d8524/)

# Supercedes/Implements

* https://github.com/StackStorm/st2/pull/6059
* https://github.com/StackStorm/st2/pull/6058
* https://github.com/StackStorm/st2/pull/6057
* https://github.com/StackStorm/st2/pull/6056
* https://github.com/StackStorm/st2/pull/6054
* https://github.com/StackStorm/st2/pull/6053

### CODE REVIEW

File: fixed-requirements.txt [hunk 1]
```diff
@@ -7,12 +7,12 @@ chardet<3.1.0
 cffi<1.15.0
 # NOTE: 2.0 version breaks pymongo work with hosts
 dnspython>=1.16.0,<2.0.0
-cryptography==39.0.1
+cryptography==41.0.4
```
Comment (armab) [inline fixed-requirements.txt:10]: `cryptography` 3.4.7 -> 39.0.1, `pyOpenSSL` 21.0.0 -> 23.1.0 and `paramiko` 2.10.5 -> 2.11.0  were just updated here https://github.com/StackStorm/st2/pull/6055 last week which fixed high and medium issues, but left one of low severity (leaving as is).

We can't update to cryptography `41` as that'll break the py3.6 support which is required for 3.8.1 patch version to be released. The team is going to drop py3.6 in the next st2 `v3.9.0` afterwards.


Please remove those changes and let's see what's left in terms of failing 🔴  tests.

---

I expect to see some difficulty with `gitpython` which also should support py3.6 somehow.

**Review by armab — CHANGES_REQUESTED** [review]
Thanks!

Left a few first pointers. The main indicator is 🟢 tests, where some more work will be needed to get them all happy for both py3.6 and 3.8+.

File: fixed-requirements.txt (hunk 1 above)
Comment (jk464) [reply fixed-requirements.txt]: Ah yes - I forgot to check that, I only tested against `py3.8` - I've dropped the changes to `cryptography` (and `pyOpenSSL` as well) now - let's see how that goes

**Comment by jk464** [comment]
@armab As you can probably see from my commits I've hit a bit of a depedency hell trying to get requirement ranges that:

* Allow the last supported py3.6
* A version with relevant CVEs fixed in py3.8

I can see in https://github.com/StackStorm/st2/pull/6063 you've hopefully got `gitpython` handled.

I'll probably do the same as you here and split this into bit size PRs to make it more manageable.

I did look at fixing `fixate-requriements.py` to support reading `python_version` which I actually got working (See https://github.com/StackStorm/st2/pull/6062/commits/61b47db603c7d21a06142228f031713ba77d7a35) (which is probably worth having, even after dropping py3.6 in `3.9.0` - issue I hit was the `stackstorm/packagingbuild:bionic` was missing the package that I wanted to using (`packaging`) so that image would also need bumped.

**Comment by armab** [comment]
@jk464 Your enhancement looks really clean. But overall I felt like env markers are buggy in many places, including older pip version we're locked to (because of py3.6) and even pants that doesn't support them in requirements-pants.txt so touching them might be like opening a can of worms.
I hope the builds would be migrated to pants from all the old machinery and so `fixate-requriements.py` may be obsolete by then.

@cognifloyd do you think it's doable to migrate to the pants builds in the upcoming v3.9.0?

**Review by armab — CHANGES_REQUESTED** [review]
With the changes extracted into a dedicated PRs one dependency at a time, let's make sure this PR updates only the following:

`requests[security]` and `importlib-metadata`

I think that's the only thing left to update for security.

File: st2client/requirements.txt [hunk 2]
```diff
@@ -21,8 +21,9 @@ python-dateutil==2.8.1
 python-editor==1.0.4
 pytz==2021.1
 pyyaml==5.4.1
-requests[security]==2.25.1
+requests[security]>=2.27.1,<=2.31.0
 six==1.13.0
 sseclient-py==1.7
 typing-extensions<4.2
+urllib3<2; python_version < '3.7'
```
Comment (armab) [inline st2client/requirements.txt:28]: Yeah, environment markers are all buggy with the current pip version 🤯 


Under py3.8 https://app.circleci.com/pipelines/github/StackStorm/st2/4267/workflows/6d50c81f-3524-4f8d-8279-71f47ee5dd67/jobs/16065?invite=true#step-109-925583_94:
```
[package: st2] [20:53:43]      The conflict is caused by:
[package: st2] [20:53:43]          requests[security] 2.31.0 depends on urllib3<3 and >=1.21.1
[package: st2] [20:53:43]          requests 2.31.0 depends on urllib3<3 and >=1.21.1
[package: st2] [20:53:43]          st2client 3.9.dev0 depends on urllib3<2
```
which is wrong.

I can't see any path forward for updating `requests[security]` so far as it requires `urllib3` which is incompatible with python3.6.

**Comment by armab** [comment]
So `importlib-metadata` changes are extracted into a dedicated PR which could be merged ASAP: https://github.com/StackStorm/st2/pull/6072

**Comment by jk464** [comment]
@armab if the build is failing for `requests` and the update to `importlib-metadata` was merged in #6072 is there anything left to do in this PR?

The only CVE I see listed against `requests` is [CVE-2023-32681](https://nvd.nist.gov/vuln/detail/CVE-2023-32681) - which is only `medium` Sev - so I think we could just drop fixing that to `3.9.0` at which point we can(?) drop python `3.6` and bump this w/o issue. (And I assume for `3.9.0` we'd probably rather being bumping dependencies to the highest support by `3.8`.)

Let me know what you think and I can close this PR if there's nothing further to do

**Comment by armab** [comment]
@jk464 Yeah, let's reassign this PR to the `v3.9.0` roadmap
It'll help keeping in mind that `requests` will need an update after dropping the py3.6.

**Review by cognifloyd — COMMENTED** [review]
This is no longer needed as we have dropped py3.6 in the master branch, and we've already bumped requests to 2.31.0.

**Pull request closed without merging**